src/aiida_pseudo/cli/install.py

```diff
@@ -143,117 +140,215 @@ def download_sssp(
+            'cannot specify both `--download-only` and `--from-download`.',
+            param_hint="'--download-only' / '--from-download'"
+        )
 
     configuration = SsspConfiguration(version, functional, protocol)
-    label = SsspFamily.format_configuration_label(configuration)
 
     if configuration not in SsspFamily.valid_configurations:
-        echo.echo_critical(f'{configuration} is not a valid configuration.')
-
-    if not download_only and QueryBuilder().append(SsspFamily, filters={'label': label}).first():
-        echo.echo_critical(f'{SsspFamily.__name__}<{label}> is already installed')
+        echo.echo_critical(f'{version} {functional} {protocol} is not a valid SSSP configuration')
 
     with tempfile.TemporaryDirectory() as dirpath:
 
         dirpath = pathlib.Path(dirpath)
+
         filepath_archive = dirpath / 'archive.tar.gz'
         filepath_metadata = dirpath / 'metadata.json'
 
-        patch_version = download_sssp(configuration, filepath_archive, filepath_metadata, traceback)
+        if from_download is not None:
 
-        description = f'SSSP v{patch_version} {functional} {protocol} installed with aiida-pseudo v{__version__}'
-        description += f'\nArchive pseudos md5: {md5_file(filepath_archive)}'
-        description += f'\nPseudo metadata md5: {md5_file(filepath_metadata)}'
+            tarball_path = pathlib.Path(from_download).absolute()
+            with tarfile.open(tarball_path, 'r') as handle:
+                handle.extractall(dirpath)
+
+            filepath_configuration = dirpath / 'configuration.json'
+
+            with filepath_configuration.open('r') as handle:
+                configuration = SsspConfiguration(**json.load(handle))
+        else:
+            download_sssp(configuration, filepath_archive, filepath_metadata, traceback)
+
+        label = SsspFamily.format_configuration_label(configuration)
+
+        if not download_only and QueryBuilder().append(SsspFamily, filters={'label': label}).first():
+            echo.echo_critical(f'{SsspFamily.__name__}<{label}> is already installed')
 
         if download_only:
-            for filepath in [filepath_archive, filepath_metadata]:
-                filepath_target = pathlib.Path.cwd() / filepath.name
-                if filepath_target.exists():
-                    echo.echo_warning(f'the file `{filepath_target}` already exists, skipping.')
-                else:
-                    # Cannot use ``pathlib.Path.rename`` because this will fail if it moves across file systems.
-                    shutil.move(filepath, filepath_target)
-                    echo.echo_success(f'`{filepath_target.name}` written to the current directory.')
+            filepath_configuration = dirpath / 'configuration.json'
+
+            with filepath_configuration.open('w') as handle:
+                handle.write(json.dumps(configuration._asdict()))
+
+            tarball_path = pathlib.Path.cwd() / f'{label}.aiida_pseudo'.replace('/', '_')
+
```

Comment:
The logic in this block is quite hard to follow. I think it will be made a lot easier if you simply take out the case for `download_only == True`. Take that out of the `with tempfile` context (or give it its own) and simply return after it is done.
```
if download_only:
    with tempfile.TemporaryDirectory() as dirpath:
         # logic to download
         return

# Here we now know we should install normally or `--from-download`
```

Reply:
Hmm, I understand how the logic is a bit tricky to follow, but moving the `download_only` up out of the `tempfile` context would lead to some code duplication. In the end the current logic is not too bad, no? Basically, in the `tempfile` context:

```
if from_download is not None:
    # Unpack and check tarball
else: (this is now an extra `if` statement so we only have to check the configuration once)
    # Download archives

if download_only:
    # Pack archives and configuration into tarball
    return

# Install pseudo family
```

Reply:
My point is that you essentially have two scenario's, `scenario A` and `scenario B`, which are completely independent and als mutually exclusive. However, the code currently flows as:
```
scenario A
scenario B
scenario A
```
This is what makes it "difficult" to follow, or "more" difficult then it should be. Why not write it as
```
scenario B
scenario A
```
So just start with the `--download-only` path and return early, and then it is clear that all that remains is for the `--from-download` and normal install path.